library/rtp/src/main/java/com/google/android/exoplayer2/source/rtp/RtpPacket.java

```diff
@@ -0,0 +1,372 @@
+ *
+ */
+public final class RtpPacket {
+    public static final int MAX_PACKET_SIZE = 65507;
+
+    private static final int RTP_VERSION = 0x02;
+
+    private static final int RTP_MIN_SIZE = 4;
+    private static final int RTP_HDR_SIZE = 12; /* RFC 3550 */
+    private static final int RTP_MAX_HDR_SIZE = 76;
+
+    private static final int RTP_XTHDR_SIZE = 4;
+
+    private static final int CSRC_SIZE = 4;
+
+    /* MPEG payload-type constants */
+    private static final int RTP_MPA_TYPE = 0x0E;     // MPEG-1 and MPEG-2 audio
+    private static final int RTP_MPV_TYPE = 0x20;     // MPEG-1 and MPEG-2 video
+
+    //Fields that compose the RTP header
+    private int version;
+    private boolean padding;
+    private boolean extension;
+    private int csrcCount;
+
+    private boolean marker;
+    private int sequenceNumber;
+    private int payloadType;
+
+    private long timestamp;
+    private long ssrc;
+
+    private long[] csrc;
+
+    private byte[] hdrExtension;
+
+    private byte[] payload;
+
+    private int extLen;
+    private int padLen;
+
+    RtpPacket(Builder builder) {
+        this.version = builder.version;
+        this.padding = builder.padding;
+        this.extension = builder.extension;
+        this.marker = builder.marker;
+        this.sequenceNumber = builder.sequenceNumber;
+        this.payloadType = builder.payloadType;
+        this.timestamp = builder.timestamp;
+        this.ssrc = builder.ssrc;
+        this.csrc = builder.csrc;
+        this.payload = builder.payload;
+
+        csrcCount = csrc != null ? csrc.length : 0;
+    }
+
+    RtpPacket(int version, int padLen, byte[] hdrExtension, boolean marker, int payloadType,
+              int sequenceNumber, long timestamp, long ssrc, long[] csrc, byte[] payload) {
+        this.version = version;
+        this.padding = padLen > 0;
+        this.padLen = padding ? padLen : 0;
+        this.extension = hdrExtension != null && hdrExtension.length > 0;
+        this.hdrExtension = hdrExtension;
+        this.marker = marker;
+        this.payloadType = payloadType;
+        this.sequenceNumber = sequenceNumber;
+        this.timestamp = timestamp;
+        this.ssrc = ssrc;
+        this.csrc = csrc;
+        this.payload = payload;
+
+        csrcCount = csrc != null ? csrc.length : 0;
+        extLen = (extension) ? hdrExtension.length : 0;
+    }
+
+    public int version() { return version; }
+
+    public boolean marker() { return marker; }
+
+    public boolean hasExtension() { return extension; }
+
+    public int payloadType() { return payloadType; }
+
+    public int sequenceNumber() { return sequenceNumber; }
+
+    public long timestamp() { return timestamp; }
+
+    public long ssrc() { return ssrc; }
+
+    public long[] csrc() { return csrc; }
+
+    public byte[] payload() { return payload; }
+
+    public byte[] extension() { return hdrExtension; }
+
+
+    public byte[] toBytes() {
```

Comment:
@tresvecesseis Hi, I found an issue with `toBytes()` method, When you are building back the complete RTP packet, the `hdrExtension` is not added, but the flag, that the packet has extension, is. When a packet with extension arrives, it is parsed and added to `RtpSamplesQueue`, after getting the packet back from queue and restoring its contents it is mising the extension and the `DefaultRtpExtractor:77` cannot parse it.

Reply:
I fixed it this way:
```
    public byte[] toBytes() {
        // build the header
        int payloadLen = (payload == null) ? 0 : payload.length;
        byte[] packet = new byte[RTP_HDR_SIZE + extLen + payloadLen];

        // fill the header array of byte with RTP header fields
        packet[0] = (byte)(version << 6 | (padding ? 1 : 0) << 5 | (extension ? 1 : 0) << 4 | csrcCount);
        packet[1] = (byte)((marker ? 1 : 0) << 7 | payloadType & 0x000000FF);
        packet[2] = (byte)(sequenceNumber >> 8);
        packet[3] = (byte)(sequenceNumber & 0xFF);
        packet[4] = (byte)(timestamp >> 24);
        packet[5] = (byte)(timestamp >> 16);
        packet[6] = (byte)(timestamp >> 8);
        packet[7] = (byte)(timestamp & 0xFF);
        packet[8] = (byte)(ssrc >> 24);
        packet[9] = (byte)(ssrc >> 16);
        packet[10] = (byte)(ssrc >> 8);
        packet[11] = (byte)(ssrc & 0xFF);

        if (extLen > 0) {
            System.arraycopy(hdrExtension, 0, packet, RTP_HDR_SIZE, extLen);
        }
        if (payloadLen > 0) {
            System.arraycopy(payload, 0, packet, RTP_HDR_SIZE + extLen, payloadLen);
        }

        return packet;
    }
```

Reply:
@jiriadamec Your fix proposal is ok. We see it in more detail and we fix it. Thanks. ;)

Reply:
@jiriadamec The same issue occurs with the csrc identifiers. Now I make a new fix with the complete solution. ;)

Reply:
ok, thank you :)

Reply:
The fix:

    public byte[] toBytes() {
        // build the header
        int payloadLen = (payload == null) ? 0 : payload.length;
        int cscrLen = csrcCount * 4;
        byte[] packet = new byte[RTP_HDR_SIZE + cscrLen + extLen + payloadLen];

        // fill the header array of byte with RTP header fields
        packet[0] = (byte)(version << 6 | (padding ? 1 : 0) << 5 | (extension ? 1 : 0) << 4 | csrcCount);
        packet[1] = (byte)((marker ? 1 : 0) << 7 | payloadType & 0x000000FF);
        packet[2] = (byte)(sequenceNumber >> 8);
        packet[3] = (byte)(sequenceNumber & 0xFF);
        packet[4] = (byte)(timestamp >> 24);
        packet[5] = (byte)(timestamp >> 16);
        packet[6] = (byte)(timestamp >> 8);
        packet[7] = (byte)(timestamp & 0xFF);
        packet[8] = (byte)(ssrc >> 24);
        packet[9] = (byte)(ssrc >> 16);
        packet[10] = (byte)(ssrc >> 8);
        packet[11] = (byte)(ssrc & 0xFF);

        // fill the contributing source identifiers
        if (csrcCount > 0) {
            for (int ndx = 0, offset = RTP_HDR_SIZE; ndx < csrcCount; ndx++, offset += 4) {
                packet[offset] = (byte) (csrc[ndx] >> 24);
                packet[offset + 1] = (byte) (csrc[ndx] >> 16);
                packet[offset + 2] = (byte) (csrc[ndx] >> 8);
                packet[offset + 3] = (byte) (csrc[ndx] & 0xFF);
            }
        }

        // fill the extension header
        if (extLen > 0) {
            System.arraycopy(hdrExtension, 0, packet, RTP_HDR_SIZE + cscrLen, extLen);
        }

        // fill the payload data
        if (payloadLen > 0) {
            System.arraycopy(payload, 0, packet, RTP_HDR_SIZE + cscrLen + extLen, payloadLen);
        }

        return packet;
    }

Reply:
👍